File: app/src/main/java/at/shockbytes/corey/fragment/WorkoutOverviewFragment.java
package at.shockbytes.corey.fragment;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.app.Fragment;
import android.support.v4.util.Pair;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import at.shockbytes.corey.R;
import at.shockbytes.corey.adapter.WorkoutAdapter;
import at.shockbytes.corey.core.CoreyApp;
import at.shockbytes.corey.core.CreateWorkoutActivity;
import at.shockbytes.corey.core.WorkoutDetailActivity;
import at.shockbytes.corey.storage.live.LiveWorkoutUpdateListener;
import at.shockbytes.corey.util.AppParams;
import at.shockbytes.corey.common.core.util.view.RecyclerViewWithEmptyView;
import at.shockbytes.corey.workout.WorkoutManager;
import at.shockbytes.corey.common.core.workout.model.Workout;
import butterknife.Bind;
import butterknife.ButterKnife;
import rx.functions.Action1;

import static com.facebook.FacebookSdk.getApplicationContext;

/**
 * @author Martin Macheiner
 *         Date: 26.10.2015.
 */
public class
WorkoutOverviewFragment extends Fragment implements WorkoutAdapter.OnItemClickListener,
        WorkoutAdapter.OnWorkoutPopupItemSelectedListener, LiveWorkoutUpdateListener {

    public static WorkoutOverviewFragment newInstance() {
        WorkoutOverviewFragment fragment = new WorkoutOverviewFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    private WorkoutAdapter adapter;

    @Inject
    protected WorkoutManager workoutManager;

    @Bind(R.id.fragment_training_rv)
    protected RecyclerViewWithEmptyView recyclerView;

    @Bind(R.id.fragment_training_empty_view)
    protected View emptyView;

    public WorkoutOverviewFragment() {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((CoreyApp) getActivity().getApplication()).getAppComponent().inject(this);
    }

    @Override
    public void onDestroy() {
        workoutManager.unregisterLiveForWorkoutUpdates(this);
        super.onDestroy();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_workout_overview, container, false);
        ButterKnife.bind(this, v);
        return v;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initialize();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ButterKnife.unbind(this);
    }

    @Override
    public void onItemClick(Workout w, View v) {

        Intent intent = WorkoutDetailActivity.newIntent(getContext(), w);
        ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(
                getActivity(),
                new Pair<>(v.findViewById(R.id.item_training_imgview_body_region),
                        getString(R.string.transition_workout_body_region)),
                new Pair<>(v.findViewById(R.id.item_training_container_duration),
                        getString(R.string.transition_workout_duration)),
                new Pair<>(v.findViewById(R.id.item_training_cardview),
                        getString(R.string.transition_workout_card)),
                new Pair<>(v.findViewById(R.id.item_training_txt_title),
                        getString(R.string.transition_workout_name)),
                new Pair<>(v.findViewById(R.id.item_training_container_exercises),
                        getString(R.string.transition_workout_exercise_count))
        );
        getActivity().startActivity(intent, options.toBundle());
    }

    @Override
    public void onDelete(Workout w) {
        adapter.deleteEntity(w);
        workoutManager.deleteWorkout(w);
    }

    @Override
    public void onEdit(Workout w) {
        Intent intent = CreateWorkoutActivity.newIntent(getApplicationContext(), w);
        ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(getActivity());
        getActivity().startActivityForResult(intent, AppParams.REQUEST_CODE_CREATE_WORKOUT, options.toBundle());
    }

    private RecyclerView.LayoutManager getLayoutManagerForOrientation() {

        if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT) {
            return new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false);
        } else {
            return new StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL);
        }
    }

    private void initialize() {

        recyclerView.setLayoutManager(getLayoutManagerForOrientation());
        adapter = new WorkoutAdapter(getActivity(), new ArrayList<Workout>(), this);
        adapter.setOnItemClickListener(this);
        recyclerView.setEmptyView(emptyView);
        recyclerView.setAdapter(adapter);

        workoutManager.getWorkouts().subscribe(new Action1<List<Workout>>() {
            @Override
            public void call(List<Workout> workouts) {
                adapter.setData(workouts);
            }
        }, new Action1<Throwable>() {
            @Override
            public void call(Throwable throwable) {
                Snackbar.make(getView(), R.string.snackbar_cannot_load_workouts, Snackbar.LENGTH_LONG).show();
                throwable.printStackTrace();
            }
        });

        workoutManager.registerLiveForWorkoutUpdates(this);
    }

    @Override
    public void onWorkoutAdded(Workout workout) {
        adapter.addEntityAtLast(workout);
    }

    @Override
    public void onWorkoutDeleted(Workout workout) {
        adapter.deleteEntity(workout);
    }

    @Override
    public void onWorkoutChanged(Workout workout) {
        adapter.updateEntity(workout);
    }
}

File: app/src/main/java/at/shockbytes/corey/ui/fragment/tab/WorkoutOverviewFragment.kt
package at.shockbytes.corey.ui.fragment.tab

import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import android.os.Bundle
import androidx.core.app.ActivityOptionsCompat
import androidx.core.util.Pair
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import android.view.View
import at.shockbytes.corey.R
import at.shockbytes.corey.ui.adapter.WorkoutAdapter
import at.shockbytes.corey.common.core.workout.model.Workout
import at.shockbytes.corey.dagger.AppComponent
import at.shockbytes.corey.ui.activity.CreateWorkoutActivity
import at.shockbytes.corey.ui.activity.WorkoutDetailActivity
import at.shockbytes.corey.util.AppParams
import at.shockbytes.corey.ui.viewmodel.WorkoutOverviewViewModel
import at.shockbytes.util.adapter.BaseAdapter
import at.shockbytes.corey.util.isPortrait
import kotlinx.android.synthetic.main.fragment_workout_overview.*
import javax.inject.Inject

/**
 * Author:  Martin Macheiner
 * Date:    26.10.2015
 */
class WorkoutOverviewFragment : TabBaseFragment<AppComponent>(),
        BaseAdapter.OnItemClickListener<Workout>,
        WorkoutAdapter.OnWorkoutPopupItemSelectedListener {

    override val snackBarBackgroundColorRes: Int = R.color.sb_background
    override val snackBarForegroundColorRes: Int = R.color.sb_foreground

    private lateinit var workoutOverviewAdapter: WorkoutAdapter
    private lateinit var viewModel: WorkoutOverviewViewModel

    override val castsActionBarShadow: Boolean = false

    @Inject
    protected lateinit var vmFactory: ViewModelProvider.Factory

    private val layoutManagerForOrientation: RecyclerView.LayoutManager
        get() = if (isPortrait()) {
            LinearLayoutManager(activity, RecyclerView.VERTICAL, false)
        } else {
            StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
        }

    override val layoutId = R.layout.fragment_workout_overview

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProviders.of(this, vmFactory)[WorkoutOverviewViewModel::class.java]
    }

    override fun injectToGraph(appComponent: AppComponent?) {
        appComponent?.inject(this)
    }

    override fun onItemClick(t: Workout, position: Int, v: View) {

        val intent = WorkoutDetailActivity.newIntent(requireContext(), t)
        val options = ActivityOptionsCompat.makeSceneTransitionAnimation(
                requireActivity(),
                Pair(v.findViewById(R.id.item_training_imgview_body_region),
                        getString(R.string.transition_workout_body_region)),
                Pair(v.findViewById(R.id.item_training_container_duration),
                        getString(R.string.transition_workout_duration)),
                Pair(v.findViewById(R.id.item_training_cardview),
                        getString(R.string.transition_workout_card)),
                Pair(v.findViewById(R.id.item_training_txt_title),
                        getString(R.string.transition_workout_name)),
                Pair(v.findViewById(R.id.item_training_container_exercises),
                        getString(R.string.transition_workout_exercise_count)),
                Pair(v.findViewById(R.id.item_training_imgview_equipment),
                        getString(R.string.transition_workout_equipment)))
        activity?.startActivity(intent, options.toBundle())
    }

    override fun onDelete(w: Workout?) {
        if (w != null) {
            // workoutOverviewAdapter.deleteEntity(w)
            viewModel.deleteWorkout(w)
        }
    }

    override fun onEdit(w: Workout?) {
        val intent = CreateWorkoutActivity.newIntent(requireContext(), w)
        val options = ActivityOptionsCompat.makeSceneTransitionAnimation(requireActivity())
        activity?.startActivityForResult(intent, AppParams.REQUEST_CODE_CREATE_WORKOUT, options.toBundle())
    }

    override fun setupViews() {

        workoutOverviewAdapter = WorkoutAdapter(
                requireContext(),
                onWorkoutPopupItemSelectedListener = this,
                onItemClickListener = this
        )

        fragment_training_rv.apply {
            layoutManager = layoutManagerForOrientation
            adapter = workoutOverviewAdapter
        }
    }

    override fun bindViewModel() {

        viewModel.getWorkouts().observe(this, { state ->

            when (state) {

                is WorkoutOverviewViewModel.RetrieveWorkoutState.Success -> {
                    workoutOverviewAdapter.data = state.workouts.toMutableList()
                }
                is WorkoutOverviewViewModel.RetrieveWorkoutState.Error -> {
                    showSnackbar(getString(R.string.snackbar_cannot_load_workouts))
                }
            }
        })
    }

    override fun unbindViewModel() {
        viewModel.getWorkouts().removeObservers(this)
    }

    companion object {

        fun newInstance(): WorkoutOverviewFragment = WorkoutOverviewFragment()
    }
}
